pingcap/tidb
# config, server: Add http status host

### What problem does this PR solve? 
There was no way to set the host for the Status listener to bind to. The --status option only accepts a port number. This could be a security problem (since a user might want to only listen for status requests on an internal network interface, etc.). It's also a usability issue, since services trying to listen on external interfaces can cause firewall issues (macOS gives an annoying popup in this case, for example).

### What is changed and how it works?
I added a --status-host option that allows a user to set the host the Status listener should bind to. It defaults to 0.0.0.0, the same as the --host option.

### Check List 

Tests 

 - Unit test
 - Integration test

I did *not* add a test to bind to different addresses to make sure the server functions as expected, because I wasn't sure how to do that in a way that wouldn't break in certain environments.

 - Manual test (add detailed scripts or steps below)

I did manually test this, binding explicitly to some IP addresses, hostnames, localhost, etc.

Code changes

 - Has exported function/method change
No
 - Has exported variable/fields change
No
 - Has interface methods change
No
 - Has persistent data change
No

This change affects the "Status" config struct, adding a StatusHost member. It also adds a `--status-host` flag to tidb-server.

Side effects

 - Possible performance regression
No
 - Increased code complexity
No
 - Breaking backward compatibility
The default server-host is 0.0.0.0, which matches the default of 0.0.0.0 for host (where TiDB listens for MySQL-protocol connections). The existing behavior is to bind to an empty interface, and I think in some environments, binding to 0.0.0.0 can mean that IPv6 connections are not allowed, but I did not observe that on my Mac. So, the change is that my patch makes the Status listener bind to "0.0.0.0" by default, while the old behavior was to find to "".

Related changes

 - Need to cherry-pick to the release branch
No.
 - Need to update the documentation
Yes, to add new flag name and config file option.
 - Need to update the `tidb-ansible` repository
Possibly, if it's desirable to allow users to set the status host as part of an Ansible deployment.
 - Need to be included in the release note
Yes.
